besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java

```diff
@@ -2352,16 +2355,46 @@ private EthNetworkConfig updateNetworkConfig(final NetworkName network) {
   }
 
   private GenesisConfigFile getGenesisConfigFile() {
-    return GenesisConfigFile.fromConfig(genesisConfig());
+    return GenesisConfigFile.fromConfigWithoutAccounts(genesisConfig());
   }
 
+  private String genesisConfigString = "";
+
   private String genesisConfig() {
-    try {
-      return Resources.toString(genesisFile.toURI().toURL(), UTF_8);
-    } catch (final IOException e) {
-      throw new ParameterException(
-          this.commandLine, String.format("Unable to load genesis URL %s.", genesisFile), e);
+    if (!genesisConfigString.isEmpty()) {
+      return genesisConfigString;
+    }
+    if (genesisStateHashCacheEnabled) {
+      // If the genesis state hash is present in the database, we can use the genesis file without
+      pluginCommonConfiguration.init(
+          dataDir(),
+          dataDir().resolve(DATABASE_PATH),
+          getDataStorageConfiguration(),
+          getMiningParameters());
+      final KeyValueStorageProvider storageProvider = keyValueStorageProvider(keyValueStorageName);
+      if (storageProvider != null) {
+        boolean isGenesisStateHashPresent;
+        try {
+          // A null pointer exception may be thrown here if the database is not initialized.
+          VariablesStorage variablesStorage = storageProvider.createVariablesStorage();
+          Optional<Hash> genesisStateHash = variablesStorage.getGenesisStateHash();
+          isGenesisStateHashPresent = genesisStateHash.isPresent();
+        } catch (Exception ignored) {
+          isGenesisStateHashPresent = false;
+        }
+        if (isGenesisStateHashPresent) {
+          genesisConfigString = JsonUtil.getJsonFromFileWithout(genesisFile, "alloc");
+        }
+      }
```

Comment:
please explain why this code is needed here now, so we can find a better organization for it

Reply:
In the original code flow, `genesis file (disk)` --(1)--> `genesis string` --(2)--> `genesis struct`.

Here, the code is positioned at step (1).

If the node has the `--genesis-state-hash-cache-enabled` parameter enabled, the value of `state root hash` will be saved in `variablesStorage`.

The intent of this code is to determine if the value of `state root hash` already exists in the node's variablesStorage, and if so, during step (1), it will directly ignore the `alloc` field in the gensis.json file.

Because if the `state root hash` value already exists, it indicates that the account information in the `genesis.json` file has already been written into the state database, and there is no need to calculate the `state root hash` again later.